src/test/java/com/aws/greengrass/lifecyclemanager/KernelLifecycleTest.java

```diff
@@ -134,6 +178,206 @@ void GIVEN_kernel_WHEN_launch_with_autostart_services_THEN_autostarts_added_as_d
+                eq(DependencyType.HARD), eq(true));
+    }
+
+    @SuppressWarnings("PMD.CloseResource")
+    @Test
+    void GIVEN_kernel_WHEN_launch_with_provisioning_plugin_THEN_plugin_methods_are_invoked()
+            throws Exception {
+
+        mockProvisioning();
+        when(mockProvisioningPlugin.updateIdentityConfiguration(any())).thenReturn(mock(ProvisionConfiguration.class));
+        EZPlugins pluginMock = mock(EZPlugins.class);
+        when(mockContext.get(EZPlugins.class)).thenReturn(pluginMock);
+        doAnswer((i) -> {
+            ImplementingClassMatchProcessor func = i.getArgument(1);
+            func.processMatch(TestDeviceProvisioningPlugin.class);
+            return null;
+        }).when(pluginMock).implementing(eq(DeviceIdentityInterface.class), any());
+
+        kernelLifecycle.launch();
+        verify(mockProvisioningPlugin, timeout(1000).times(1)).updateIdentityConfiguration(any(ProvisionContext.class));
+    }
+
+
+
+    @SuppressWarnings("PMD.CloseResource")
+    @Test
+    void GIVEN_kernel_WHEN_launch_with_provisioning_plugin_THEN_configuration_is_updated()
+            throws Exception {
+
+        mockProvisioning();
+        ProvisionConfiguration mockProvisionConfiguration = createMockProvisioningConforguration();
+        when(mockProvisioningPlugin.updateIdentityConfiguration(any())).thenReturn(mockProvisionConfiguration);
+        EZPlugins pluginMock = mock(EZPlugins.class);
+        when(mockContext.get(EZPlugins.class)).thenReturn(pluginMock);
+        doAnswer((i) -> {
+            ImplementingClassMatchProcessor func = i.getArgument(1);
+            func.processMatch(TestDeviceProvisioningPlugin.class);
+            return null;
+        }).when(pluginMock).implementing(eq(DeviceIdentityInterface.class), any());
+
+        kernelLifecycle.launch();
+
+        ArgumentCaptor<NucleusConfiguration> nucleusConfigCaptor =
+                ArgumentCaptor.forClass(NucleusConfiguration.class);
+        verify(mockProvisioningHelper, timeout(500).times(1))
+                .updateNucleusConfiguration(nucleusConfigCaptor.capture()
+                        , eq(UpdateBehaviorTree.UpdateBehavior.MERGE));
+
+        ArgumentCaptor<SystemConfiguration> systemConfigCaptor =
+                ArgumentCaptor.forClass(SystemConfiguration.class);
+        verify(mockProvisioningHelper, timeout(500).times(1))
+                .updateSystemConfiguration(systemConfigCaptor.capture()
+                        , eq(UpdateBehaviorTree.UpdateBehavior.MERGE));
+        assertEquals(mockProvisionConfiguration.getNucleusConfiguration(), nucleusConfigCaptor.getValue());
+        assertEquals(mockProvisionConfiguration.getSystemConfiguration(), systemConfigCaptor.getValue());
+
+    }
+
+
+    @SuppressWarnings("PMD.CloseResource")
+    @Test
+    void GIVEN_kernel_WHEN_launch_with_provisioning_plugin_AND_plugin_methods_throw_runtime_Exception_THEN_offline_mode(ExtensionContext context)
```

Comment:
I'm still concerned with this behavior, in certain cases this might make sense but more commonly, I think it'll be better to fail the nucleus launch so that the user knows clearly there's a problem and can take steps to fix it, otherwise, the user is not going to know without looking at the logs. We'd talked earlier about adding something like an `onFailure` callback to the provisioning plugin so the plugin author gets some flexibility, where did that land?

Reply:
I do not think we should fail the launch. Nucleus should still operate in offline mode. Even if we fail nucleus launch, they will have to see logs to find out the problem. And if it is setup as systemd service it will keep restarting with the same issue. The plugin can throw retryable exceptions if it wants to retry otherwise it can fail with a non-retryable error. This should give it enough flexibility to plugin author. We do not want to give nucleus's launch control to plugin.

Reply:
System service should not be setup before provisioning is finished. Retrying is only helpful for intermittent problems. My point is mainly considering that customers writing a plugin and capable of log diving can be different from end users who may not be trained. So I think any failures in critical steps like this should be clear enough. Can you explain a bit what we gain by nucleus operating offline since it won't be able to connect to cloud anyway? I don't see that failing the launch gives any unnecessary control to the plugin, but instead makes the user's life easy. We can discuss offline if you want

Reply:
1. If we want to wait for provisioning to finish, to setup system service later, then that means provisioning needs to be synchronous or easy setup code (or nucleus) needs to wait on its completion. This is one thing we were trying to avoid during our design discussions.
2. I am going to switch calling it non-provisioned mode (from offline mode). Some components may just run in non-provisioned mode. More than non-provisioned mode, it is important to keep greengrass running so it can react to any events (local in case of offline). If nucleus waits on plugin, and the plugin for some reason hangs, this will hang nucleus for that much time (and stop other components from starting). If we put a max timeout on plugin response, we are putting a hard limit on plugin which may not be desirable and takes away the flexibility. 
3. I understand your point and I think we may need a mechanism for plugin to indicate to shutdown on failure. But even in that case if system service is setup then this will be happening again and again.

Reply:
Discussed offline